lib/logflare_web/controllers/source_controller.ex

```diff
@@ -222,47 +222,60 @@ defmodule LogflareWeb.SourceController do
   end
 
   def test_alerts(conn, %{"id" => source_id}) do
-    source = Sources.get_by_and_preload(id: source_id)
-
-    case WebhookNotificationServer.test_post(source) do
-      {:ok, %Tesla.Env{} = _response} ->
-        conn
-        |> put_flash(:info, "Webhook test successful!")
-        |> redirect(to: Routes.source_path(conn, :edit, source.id))
-
-      {:error, %Tesla.Env{} = response} ->
-        conn
-        |> put_flash(:error, "Webhook test failed! Response status code was #{response.status}.")
-        |> redirect(to: Routes.source_path(conn, :edit, source.id))
-
-      {:error, response} ->
+    with source = %Source{} <- Sources.get_by_and_preload(id: source_id) do
+      case WebhookNotificationServer.test_post(source) do
+        {:ok, %Tesla.Env{} = _response} ->
+          conn
+          |> put_flash(:info, "Webhook test successful!")
+          |> redirect(to: Routes.source_path(conn, :edit, source.id))
+
+        {:error, %Tesla.Env{} = response} ->
+          conn
+          |> put_flash(
+            :error,
+            "Webhook test failed! Response status code was #{response.status}."
+          )
+          |> redirect(to: Routes.source_path(conn, :edit, source.id))
+
+        {:error, response} ->
+          conn
+          |> put_flash(:error, "Webhook test failed! Error response: #{response}")
+          |> redirect(to: Routes.source_path(conn, :edit, source.id))
+      end
+    else
+      nil ->
         conn
-        |> put_flash(:error, "Webhook test failed! Error response: #{response}")
-        |> redirect(to: Routes.source_path(conn, :edit, source.id))
+        |> put_flash(:error, "Source not found")
+        |> redirect(to: Routes.source_path(conn, :index))
     end
   end
 
   def test_slack_hook(conn, %{"id" => source_id}) do
-    source = Sources.get_by_and_preload(id: source_id)
-
-    case SlackHookServer.test_post(source) do
-      {:ok, %Tesla.Env{} = _response} ->
-        conn
-        |> put_flash(:info, "Slack hook test successful!")
-        |> redirect(to: Routes.source_path(conn, :edit, source.id))
-
-      {:error, %Tesla.Env{} = response} ->
-        conn
-        |> put_flash(
-          :error,
-          "Slack hook test failed! Response status code was #{response.status}."
-        )
-        |> redirect(to: Routes.source_path(conn, :edit, source.id))
-
-      {:error, response} ->
+    with source = %Source{} <- Sources.get_by_and_preload(id: source_id) do
+      case SlackHookServer.test_post(source) do
+        {:ok, %Tesla.Env{} = _response} ->
+          conn
+          |> put_flash(:info, "Slack hook test successful!")
+          |> redirect(to: Routes.source_path(conn, :edit, source.id))
+
+        {:error, %Tesla.Env{} = response} ->
```

Comment:
These matches can be in the else instead

Reply:
Done in last refactor :bow: